### PR TITLE
Backport fix gwb clang flags.

### PR DESCRIPTION
**Pull request by MFraters**
This is a backport of https://github.com/GeodynamicWorldBuilder/WorldBuilder/pull/739, which applies the fixes for the stricter clang flags. I did this by hand more or less, so hopefully I didn't miss anything. 

@tjhei can you run this with your clang aspect version?

### CODE REVIEW

File: contrib/world_builder/include/world_builder/bounding_box.h [hunk 1]
```diff
@@ -93,7 +93,7 @@ namespace WorldBuilder
    *
    * @note The majority of this class is copied from the deal.II library.
    */
-  template <int spacedim>
+  template <unsigned int spacedim>
```
Comment (bangerth) [inline contrib/world_builder/include/world_builder/bounding_box.h:96]: Wait, you really did make that change? Maybe I need to reconsider #5844 :-)

Comment (MFraters) [reply contrib/world_builder/include/world_builder/bounding_box.h:96]: It is the easiest way to get rid of the new warnings when using clang/the new gcc flag I added in he world builder and logically the most consistent. I don't think it will impact ASPECT in any way, since the types are not shared. For aspect it may be more difficult since a lot is shared with dealii, but maybe not.

**Comment by MFraters** [comment]
I have addressed @tjhei latests comment in https://github.com/GeodynamicWorldBuilder/WorldBuilder/pull/739 here as well.

**Comment by MFraters** [comment]
@gassmoeller, I fixed the issue you found with the extra space before the template

File: contrib/world_builder/include/world_builder/features/mantle_layer_models/composition/interface.h [hunk 2]
```diff
@@ -29,7 +29,7 @@ namespace WorldBuilder
 {
   class World;
   class Parameters;
-  template <int dim> class Point;
+    template <unsigned int dim> class Point;
```
Comment (gassmoeller) [inline contrib/world_builder/include/world_builder/features/mantle_layer_models/composition/interface.h:32]: Still here in some places

File: contrib/world_builder/include/world_builder/types/int.h [hunk 3]
```diff
@@ -60,8 +60,8 @@ namespace WorldBuilder
                           const std::string &name,
                           const std::string &documentation) const override final;
 
-        unsigned int value {0};
-        unsigned int default_value;
+        int value {0};
+        int default_value;
```
Comment (gassmoeller) [inline contrib/world_builder/include/world_builder/types/int.h:64]: Why this change? What doe these values mean?

File: contrib/world_builder/source/world_builder/utilities.cc [hunk 4]
```diff
@@ -1219,7 +1217,9 @@ namespace WorldBuilder
                 }
 
               data_string = datastream.str();
-              filesize = data_string.size();
+              WBAssertThrow(static_cast<long int>(data_string.size()) < std::numeric_limits<int>::max(),
+                            "File to large to be send with MPI.");
```
Comment (gassmoeller) [inline contrib/world_builder/source/world_builder/utilities.cc:1221]:
```suggestion
                            "File is too large to be send with MPI.");
```

**Review by gassmoeller — COMMENTED** [review]
Just small comments left.

File: contrib/world_builder/include/world_builder/types/int.h (hunk 3 above)
Comment (MFraters) [reply contrib/world_builder/include/world_builder/types/int.h:64]: This was a bug, it should always have been int, because this is a type which stores a int. The values store the actual and the default value in the type.

**Comment by MFraters** [comment]
@gassmoeller, sorry that I missed a few. I think I got them all now, and I have addressed your other comments.

**Comment by tjhei** [comment]
```
/ssd/aspect-git/contrib/world_builder/source/world_builder/utilities.cc:1193:48: warning: implicit conversion changes signedness: 'int' to 'unsigned int' [-Wsign-conversion]
 1193 |               unsigned int invalid_file_size = -1;
 ```

**Review by tjhei — APPROVED** [review]
